src/rand.rs

```diff
@@ -33,6 +33,73 @@ pub trait SecureRandom: sealed::Sealed {
     fn fill(&self, dest: &mut [u8]) -> Result<(), error::Unspecified>;
 }
 
+/// A trait for modifying the underlying bytes of a type, used by `ring::rand` for random initialization.
+pub trait AsBytesMut {
+    /// Access the raw bytes underlying `self`.
+    fn as_bytes_mut(&mut self) -> &mut [u8];
+}
+
+impl AsBytesMut for u8 {
```

Comment:
A lot of these implementations can be made a lot clearer using generic implementations and better macros. We should only have to implement AsBytesMut for all of the integer types `u8`, `i8`, etc... and then just have generic implementations `impl<T:AsBytesMut> AsBytesMut for [T; $N]`. This make the implementation shorter and also allowing it to work with types like `[[u16; 2]; 2]`.

Here is [my implementation idea](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=454d2d0392e5f7d69123500e27a861a0) for `AsBytesMut`.

Reply:
Actually, the above implementation is unsafe. Not having any of https://github.com/rust-lang/rust/issues/43408, https://github.com/rust-lang/rust/issues/44580, or https://github.com/rust-lang/rust/issues/31844 makes it hard to implement this in a nice way. To make it better we could do one of the following:

If we just want to solve #728, it would be enough to define two methods ([code link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=67e7cca8e0d2248ca9494ce8872be235)):
```rust
fn bytes<B: AsMut<[u8]> + Default>(&self) -> Result<B, Unspecified>;
fn value<V: FromBytes>(&self) -> Result<V, Unspecified>;
```
Then a user could define `FromBytes` however they wanted. We would just implement `FromBytes` for the primitive integer types. This also allows for an implementation with no `unsafe` code. However, then things like `[usize; 2]` wouldn't work.

Otherwise, we could define two traits:
  - an unsafe trait `IsBytes` that essentially says "this type can be `mem::transmute`d from a random byte slice"
  - a safe trait `AsBytesMut` that is implemented for all `IsBytes` types.
Then [this implementation](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e3b470688ed20d83721aa7e0d3873422) would work, and users could define their own types to be `AsBytesMut`. We would probably want to keep `IsBytes` sealed. This requires having two traits in the public API (and using `unsafe`), but then we can have a single method:
```rust
fn value<T: AsBytesMut + Default>(&self) -> Result<T, error::Unspecified>;
```
A variant of the above could just use a `FromRand` trait instead of a `value` method
```rust
pub trait FromRand {
    fn from_rand(rng: &impl SecureRandom) -> Result<Self, Unspecified>;
}
```
and have an implementation [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b31145a5ce33a051ed129ee8d4d13f1c), but the concerns are quite similar.